### DISCUSSION TITLE
Triggers runs correctly when row is updated from dashboard but not from js client

### DISCUSSION THREAD

**Discussion by philbaker4** [supabase/supabase#4751] [Questions] [answered]
Running into a weird issue, figured it might be best to start as a discussion prior to opening an issue.. 

I have a trigger that runs AFTER UPDATE on table A. Inside this trigger, I update table B and table C. 

When I manually updated table A using the table editor UI or via a SQL query, the trigger runs correctly. 

When I update table A using the JS client as shown below, table A is updated correctly, the trigger updates table B correctly, but table C is untouched for some reason.. 

```js
await supabaseClient
      .from('A')
      .update(data)
      .match({ id: data.id })
```
My understanding was that all of this was happening at the postgres level and I really am struggling to understand why the method of updating table A would at all impact the trigger function.. 

Anybody have any thoughts or resources for me?

**Comment by GaryAustin1** [comment, marked as answer]
Do you have RLS on table C?  When you trigger off of a client update your function runs with the clients privileges.  
You can use "security definer" in your function to have it execute with the function owners privileges.
https://supabase.com/docs/guides/auth/row-level-security#policies-with-security-definer-functions

**Reply by philbaker4** [reply]
Ah, great thought. It definitely feels RLS related. I was thinking that because the function was running via after update trigger, RLS would be bypassed. Will the trigger function have access to the auth context though? From some basic testing, it seems as though the answer is no.. Do you think I'll need to open up update RLS for that table and handle security at the API level?

**Reply by GaryAustin1** [reply]
Using security definer in the function ignores RLS if created in the UI or by postgres or higher user.
I  would not think RLS on your table A has anything to do with it as it works in all cases.  I'm presuming since you don't show any code or RLS rules it is specific to table C update/insert.  You now mention auth context, is table B or C in the auth schema?  
It is possible to have a path issue also if the tables are not in public and you don't specify their schema with schema.table in the function if the client does not have the proper paths set. 
If you need to do any security you could do in the function, but client call level is not the place to deal with security.
EDIT:  The trigger function should have access to auth.uid(), etc.  It is running in the same session as the first update of table A.

**Reply by philbaker4** [reply]
Just did some testing and playing around RLS and seeing some weird results.. 

If I remove all RLS, everything works correctly. If I add RLS back to table C alone with a policy that simply returns true for updates, the trigger issue returns - trigger function is unable to update table C via js client.

Seemingly related to RLS, but unsure why this policy would cause this issue within a trigger function
![image](https://user-images.githubusercontent.com/17916326/147833204-9e60ee32-76b8-44f7-be65-cdf26aaa842e.png)

**Reply by GaryAustin1** [reply]
You might show your function.  You could show your actual RLS also.  Update needs at least "using" with optional "with check".  Also add security definer to the function to verify it ignores that RLS on C.

**Reply by philbaker4** [reply]
![image](https://user-images.githubusercontent.com/17916326/147833821-fe7c0b51-7597-4f6a-b4dc-0e9fd11906cc.png)

Not sure what you mean about the security definer portion...

**Reply by GaryAustin1** [reply]
https://supabase.com/docs/guides/auth/row-level-security#policies-with-security-definer-functions shows an example of using it in a RLS function, but same applies to your trigger function.

If you add "security definer" to the function it will run like the user who created the function (and bypass RLS in most cases).

Are you sure you have no other RLS on, or do more than updates in your function (select,insert)?  Your update RLS should work for client so it is likely some other RLS getting you.

**Reply by philbaker4** [reply]
ah I was completely misunderstanding your suggestion. Got it working now, thanks so much
